worker/download.go

```diff
@@ -0,0 +1,1088 @@
+	return len(mgr.ongoing)
+}
+
+func (mgr *downloadManager) downloadSlab(ctx context.Context, dID id, slice object.SlabSlice, index int, responseChan chan *slabDownloadResponse, nextSlabChan chan struct{}) {
+	// add tracing
+	ctx, span := tracing.Tracer.Start(ctx, "downloadSlab")
+	defer span.End()
+
+	// prepare the download
+	slab, finishFn := mgr.newSlabDownload(ctx, dID, slice, index)
+	defer finishFn()
+
+	// download shards
+	resp := &slabDownloadResponse{index: index}
+	resp.shards, resp.err = slab.downloadShards(ctx, nextSlabChan)
+
+	// send the response
+	select {
+	case <-ctx.Done():
+	case responseChan <- resp:
+	}
+}
+
+func (d *downloader) stats() downloaderStats {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	return downloaderStats{
+		avgSpeedMBPS: d.statsDownloadSpeedBytesPerMS.Average() * 0.008,
+		healthy:      d.consecutiveFailures == 0,
+		numDownloads: d.numDownloads,
+	}
+}
+
+func (d *downloader) isStopped() bool {
+	select {
+	case <-d.stopChan:
+		return true
+	default:
+	}
+	return false
+}
+
+func (d *downloader) processQueue(hp hostProvider) {
+outer:
+	for {
+		// wait for work
+		select {
+		case <-d.signalWorkChan:
+		case <-d.stopChan:
+			return
+		}
+
+		for {
+			// check if we are stopped
+			if d.isStopped() {
+				return
+			}
+
+			// fill batch
+			var batch []*sectorDownloadReq
+			for len(batch) < maxConcurrentSectorsPerHost {
+				if req := d.pop(); req == nil {
+					break
+				} else if req.done() {
+					continue
+				} else {
+					batch = append(batch, req)
+				}
+			}
+			if len(batch) == 0 {
+				continue outer
+			}
+
+			fmt.Printf("DEBUG PJ: downloader %v is processing batch of size %v\n", d.host.HostKey(), len(batch))
+
+			var mu sync.Mutex
+			var completed int
+			var failed int
+			var first time.Time
+
+			// launch workers
+			var wg sync.WaitGroup
+			jobsChan := make(chan *sectorDownloadReq)
+			for i := 0; i < len(batch); i++ {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					for job := range jobsChan {
+						// check if we are stopped
+						if d.isStopped() {
+							return
+						}
+
+						// execute the job
+						start := time.Now()
+						err := d.execute(job)
+						d.trackFailure(err)
+
+						// update stats
+						mu.Lock()
+						if first.IsZero() {
+							first = start
+						}
+						if err != nil {
+							failed++
+							completed = 0
+							d.statsSectorDownloadEstimateInMS.Track(float64(time.Hour.Milliseconds()))
+							fmt.Printf("DEBUG PJ: host %v errored out\n", d.host.HostKey())
+						} else {
+							completed++
+							if completed == maxConcurrentSectorsPerHost {
+								downloadedB := int64(completed * rhpv2.SectorSize)
+								durationMS := time.Since(first).Milliseconds()
+								d.statsDownloadSpeedBytesPerMS.Track(float64(downloadedB / durationMS))
+								d.statsSectorDownloadEstimateInMS.Track(float64(durationMS))
+								completed = 0
+								first = time.Time{}
+							}
+						}
+						mu.Unlock()
+					}
+
+					// update stats
```

Comment:
What is this about? Didn't you just update the stats in the for loop?

Reply:
I'll try and clean this up, but the idea is that once I've spun up a thread that uploads to a single host, I want to keep it going if there's new jobs coming in. So for as long as there is work, I want to keep the goroutine. I do however want to update stats from time to time, so I do it twice, once in the loop and once after the loop to "flush" the remaining stats.